Userland/Libraries/LibJS/Runtime/Value.cpp

```diff
@@ -517,6 +500,96 @@ ThrowCompletionOr<Value> Value::to_number(VM& vm) const
     }
 }
 
+constexpr bool is_ascii_number(u32 code_point)
+{
+    return is_ascii_digit(code_point) || code_point == '.' || (code_point == 'e' || code_point == 'E') || code_point == '+' || code_point == '-';
+}
+
+struct NumberParseResult {
+    StringView literal;
+    u8 base;
+};
+
+static Optional<NumberParseResult> parse_number_text(StringView text)
+{
+    NumberParseResult result {};
+
+    auto check_prefix = [&](auto lower_prefix, auto upper_prefix) {
+        if (text.length() <= 2)
+            return false;
+        if (!text.starts_with(lower_prefix) && !text.starts_with(upper_prefix))
+            return false;
+        return true;
+    };
+
+    // https://tc39.es/ecma262/#sec-tonumber-applied-to-the-string-type
+    if (check_prefix("0b"sv, "0B"sv)) {
+        if (!all_of(text.substring_view(2), is_ascii_binary_digit))
+            return {};
+
+        result.literal = text.substring_view(2);
+        result.base = 2;
+    } else if (check_prefix("0o"sv, "0O"sv)) {
+        if (!all_of(text.substring_view(2), is_ascii_octal_digit))
+            return {};
+
+        result.literal = text.substring_view(2);
+        result.base = 8;
+    } else if (check_prefix("0x"sv, "0X"sv)) {
+        if (!all_of(text.substring_view(2), is_ascii_hex_digit))
+            return {};
+
+        result.literal = text.substring_view(2);
+        result.base = 16;
+    } else {
+        if (!all_of(text, is_ascii_number))
+            return {};
+
+        result.literal = text;
+        result.base = 10;
+    }
+
+    return result;
+}
+
+// 7.1.4.1.1 StringToNumber ( str ), https://tc39.es/ecma262/#sec-stringtonumber
+Optional<Value> Value::string_to_number() const
```

Comment:
Please make this a standalone function taking a StringView, no need for this to be a `Value` method

Reply:
I see that `string_to_bigint` already uses this pattern, but let's not repeat this weirdness